File: kotlin-node/karakum/plugins/convertQuerystringAliases.js
import ts from "typescript";

export default function (node) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("querystring.d.ts")
        && ts.isVariableDeclaration(node)
        && ts.isIdentifier(node.name)
        && (
            node.name.text === "encode"
            || node.name.text === "decode"
        )
    ) {
        // ignore querystring aliases
        return ""
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertQuerystringAliases.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isIdentifier
import typescript.isVariableDeclaration

val convertQuerystringAliases = createPlugin { node, _, _ ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("querystring.d.ts"))

        ensure(isVariableDeclaration(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(
            name.text == "encode"
                    || name.text == "decode"
        )

        // ignore querystring aliases
        ""
    }
}
